<launch>

  <!-- defines name of the namespace -->
  <arg name="GS_NS" default="$(optenv GS_NS gs)" />

  <arg name="custom_config" default="" />

  <!-- will it run using GNU debugger? -->
  <arg name="DEBUG" default="false" />
  <arg unless="$(arg DEBUG)" name="launch_prefix_debug" value=""/>
  <arg     if="$(arg DEBUG)" name="launch_prefix_debug" value="debug_roslaunch"/>

  <!-- will it run as standalone nodelet or using a nodelet manager? -->
  <arg name="standalone" default="true" />
  <arg name="manager" default="$(arg GS_NS)_iroc_fleet_manager_manager" />
  <arg unless="$(arg standalone)" name="nodelet" value="load"/>
  <arg     if="$(arg standalone)" name="nodelet" value="standalone"/>
  <arg unless="$(arg standalone)" name="nodelet_manager" value="$(arg manager)"/>
  <arg     if="$(arg standalone)" name="nodelet_manager" value=""/>

  <group ns="$(arg GS_NS)">

    <!-- IROCFleetManager nodelet -->
    <node pkg="nodelet" type="nodelet" name="iroc_fleet_manager" args="$(arg nodelet) iroc_fleet_manager/IROCFleetManager $(arg nodelet_manager)" launch-prefix="$(arg launch_prefix_debug)" output="screen">

      <param name="config" value="$(find iroc_fleet_manager)/config/config.yaml" />

      <param if="$(eval arg('custom_config') == '')" name="custom_config" value="" />
      <param if="$(eval not arg('custom_config') == '' and arg('custom_config')[0] == '/')" name="custom_config" value="$(arg custom_config)" />
      <param if="$(eval not arg('custom_config') == '' and not arg('custom_config')[0] == '/')" name="custom_config" value="$(env PWD)/$(arg custom_config)" />

      <!-- Subscribers -->
      <!-- Service clients -->
      <remap from="~svc/mission_activation" to="/mission_handler/mission_activation" />
      <remap from="~svc/mission_pausing" to="/mission_handler/mission_pausing" />
      <!-- Service servers -->
      <remap from="~svc/change_fleet_mission_state" to="~change_fleet_mission_state" />
      <remap from="~svc/change_robot_mission_state" to="~change_robot_mission_state" />
      <!-- Action client -->
      <remap from="~ac/waypoint_mission" to="/mission_handler" />

    </node>

  <!-- AutonomyTest nodelet -->
    <node pkg="nodelet" type="nodelet" name="iroc_autonomy_test" args="$(arg nodelet) iroc_fleet_manager/IROCAutonomyTestManager $(arg nodelet_manager)" launch-prefix="$(arg launch_prefix_debug)" output="screen">

      <param name="config" value="$(find iroc_fleet_manager)/config/config.yaml" />

      <param if="$(eval arg('custom_config') == '')" name="custom_config" value="" />
      <param if="$(eval not arg('custom_config') == '' and arg('custom_config')[0] == '/')" name="custom_config" value="$(arg custom_config)" />
      <param if="$(eval not arg('custom_config') == '' and not arg('custom_config')[0] == '/')" name="custom_config" value="$(env PWD)/$(arg custom_config)" />

      <!-- Subscribers -->
      <!-- Service clients -->
      <remap from="~svc/mission_activation" to="/mission_handler/mission_activation" />
      <remap from="~svc/mission_pausing" to="/mission_handler/mission_pausing" />
      <!-- Service servers -->
      <remap from="~svc/change_fleet_mission_state" to="~change_fleet_mission_state" />
      <remap from="~svc/change_robot_mission_state" to="~change_robot_mission_state" />
      <!-- Action client -->
      <remap from="~ac/waypoint_mission" to="/mission_handler" />

    </node>

  <!-- CoveragePlanner nodelet -->
    <node pkg="nodelet" type="nodelet" name="iroc_coverage_planner" args="$(arg nodelet) iroc_fleet_manager/IROC_CoverageManager $(arg nodelet_manager)" launch-prefix="$(arg launch_prefix_debug)" output="screen">

      <param name="config" value="$(find iroc_fleet_manager)/config/config.yaml" />

      <param if="$(eval arg('custom_config') == '')" name="custom_config" value="" />
      <param if="$(eval not arg('custom_config') == '' and arg('custom_config')[0] == '/')" name="custom_config" value="$(arg custom_config)" />
      <param if="$(eval not arg('custom_config') == '' and not arg('custom_config')[0] == '/')" name="custom_config" value="$(env PWD)/$(arg custom_config)" />

      <!-- Subscribers -->
      <!-- Service clients -->
      <remap from="~svc/mission_activation" to="/mission_handler/mission_activation" />
      <remap from="~svc/mission_pausing" to="/mission_handler/mission_pausing" />
      <!-- Service servers -->
      <remap from="~svc/change_fleet_mission_state" to="~change_fleet_mission_state" />
      <remap from="~svc/change_robot_mission_state" to="~change_robot_mission_state" />
      <!-- Action client -->
      <remap from="~ac/waypoint_mission" to="/mission_handler" />

    </node>

  </group>

</launch>
